dev/src/serializer.ts

```diff
@@ -312,7 +312,7 @@ export function validateUserInput(
   level?: number,
   inArray?: boolean
 ): void {
-  if (path && path.size > MAX_DEPTH) {
+  if (path && path.size - 1 > MAX_DEPTH) {
```

Comment:
this may be the simplest workaround for the issue but may not be the best fix. 

IMO, L348 is the one that's problematic, it adds the path of a top-level map property to the accumulating path, creating the off-by-one error.  "foo" in `{"foo": {"bar": "baz"}}` have depth 0 but this implementation currently calculates it as depth 1. 

Currently, `path` serves two purposes: 

1. to be used in the fieldPathMessage (as part of error message)
2. to be used for counting the depth

I think we should separate this responsibility out. 

Separately, I think `validateUserInput` may be exposing too wide of an API. Currently, it can be used to validate an array property or validate a Firestore value as a whole. The ambiguity really comes from this. This also lead to incorrect usage, e.g., in https://github.com/googleapis/nodejs-firestore/blob/ce9abfd8cd65d0d2be4dd7f6a56fb679a100296a/dev/src/field-value.ts#L571, the path should include the array index in order for the depth calculation to be correct. 

I don't have a strong opinion on this though. This change is technically relaxing the requirement on the client side. If anything, we always have the backend side as fallback so the change as is may be fine.

Reply:
Thanks for the feedback.

Regarding L348 being problematic. If a path of `['foo', 'bar']` is considered to have a depth of `1` and  `['foo']` is considered to have a depth of `0`, then yes we could remove the top level map property or we could compute the depth as `path.size - 1`. It's the same IMO and allows the re-use of path.

I agree with your statement "I think validateUserInput may be exposing too wide of an API", but in consideration of our other discussion of removing all of this validation. I'm going to avoid any refactor for now.